LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceFactory.cs

```diff
@@ -73,13 +74,32 @@ private async Task<LoRaDevice> RegisterCoreAsync(IoTHubDeviceInfo deviceInfo, Ca
                 {
                     throw new LoRaProcessingException("Failed to initialize device twins.", LoRaProcessingErrorCode.DeviceInitializationFailed);
                 }
+
                 this.loRaDeviceCache.Register(loRaDevice);
+
                 return loRaDevice;
             }
             catch
             {
-                this.connectionManager.Release(loRaDevice);
-                loRaDevice.Dispose();
+                // release the loradevice client explicitly. If we were unable to register, or there was already
+                // a connection registered, we will leak this client.
+                await loRaDeviceClient.DisposeAsync();
+
+                if (this.logger.IsEnabled(LogLevel.Debug))
+                {
+                    try
+                    {
+                        // if the created client is registered, release it
+                        if (!ReferenceEquals(loRaDeviceClient, ((IIdentityProvider<ILoRaDeviceClient>)this.connectionManager.GetClient(loRaDevice)).Identity))
+                        {
+                            this.logger.LogDebug("leaked connection found");
+                        }
+                    }
+                    catch (ManagedConnectionException) { }
+                }
+
+                await this.connectionManager.ReleaseAsync(loRaDevice);
```

Comment:
nit: Isn't this redundant with the next line, where the device is disposed and which eventually ends up doing the same?

Reply:
@p-schuler, I agree with Atif. Did you have anything particular in mind here? Can we get rid of the additional ReleaseAsync?

Reply:
you can remove it - I believe this was a bit historical from how this developed (having async support on the close connection first and not the LoRaDevice.Dispose)